File: app/src/main/java/com/goloveschenko/calculator/activity/HistoryActivity.java
package com.goloveschenko.calculator.activity;

import android.app.DatePickerDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.res.Resources;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.InputType;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.DatePicker;
import android.widget.EditText;

import com.goloveschenko.calculator.tools.HistoryAdapter;
import com.goloveschenko.calculator.R;
import com.goloveschenko.calculator.dao.manager.DBManager;
import com.goloveschenko.calculator.entity.HistoryItem;
import com.goloveschenko.calculator.tools.RecyclerItemClickListener;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.List;
import java.util.Locale;

public class HistoryActivity extends AppCompatActivity {
    public static final String DATE_FORMAT = "dd MMM yyyy";
    public static final String EXTRA_EXPRESSION_RESULT = "result";

    public static final int MENU_INPUT_RESULT = 0;
    public static final int MENU_INPUT_EXPRESSION = 1;
    public static final int MENU_COMMENT = 0;
    public static final int MENU_DELETE = 1;

    private RecyclerView historyView;
    private RecyclerView.Adapter adapter;
    private List<HistoryItem> historyItems;

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.history_menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_search:
                searchItems();
                return true;
            case R.id.menu_clear:
                clearItems();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_history);

        historyView = (RecyclerView) findViewById(R.id.historyView);
        historyView.setLayoutManager(new LinearLayoutManager(this));

        historyItems = DBManager.getInstance(historyView.getContext()).selectValues();
        adapter = new HistoryAdapter(historyItems);
        historyView.setAdapter(adapter);

        historyView.addOnItemTouchListener(new RecyclerItemClickListener(historyView.getContext(), historyView, new RecyclerItemClickListener.OnItemClickListener() {
            @Override
            public void onItemClick(View view, int position) {
                Resources res = getResources();
                String[] menuItems = res.getStringArray(R.array.history_item_input_menu);
                final HistoryItem item = historyItems.get(position);
                AlertDialog.Builder menuBuilder = new AlertDialog.Builder(view.getContext());
                menuBuilder.setTitle(R.string.history_input_title);
                menuBuilder.setItems(menuItems, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        String result = "";
                        switch (which) {
                            case MENU_INPUT_RESULT:
                                result = item.getResult();
                                break;
                            case MENU_INPUT_EXPRESSION:
                                result = item.getExpression();
                                break;
                        }
                        Intent intent = new Intent();
                        intent.putExtra(EXTRA_EXPRESSION_RESULT, result);
                        setResult(RESULT_OK, intent);
                        finish();
                    }
                });
                AlertDialog alert = menuBuilder.create();
                alert.show();
            }

            @Override
            public void onLongItemClick(View view, final int position) {
                Resources res = getResources();
                String[] menuItems = res.getStringArray(R.array.history_item_menu);
                AlertDialog.Builder menuBuilder = new AlertDialog.Builder(view.getContext());
                menuBuilder.setTitle(R.string.history_item_alert_title);
                menuBuilder.setItems(menuItems, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        switch (which) {
                            case MENU_COMMENT:
                                AlertDialog.Builder commentBuilder = new AlertDialog.Builder(historyView.getContext());
                                final EditText input = new EditText(historyView.getContext());
                                input.setInputType(InputType.TYPE_CLASS_TEXT);
                                commentBuilder
                                        .setCancelable(false)
                                        .setTitle(R.string.input_comment_title)
                                        .setView(input)
                                        .setNegativeButton(android.R.string.cancel, new DialogInterface.OnClickListener() {
                                            @Override
                                            public void onClick(DialogInterface dialog, int which) {
                                                dialog.cancel();
                                            }
                                        })
                                        .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                                            @Override
                                            public void onClick(DialogInterface dialog, int which) {
                                                HistoryItem item = historyItems.get(position);
                                                String comment = input.getText().toString();
                                                if (comment.isEmpty()) {
                                                    comment = null;
                                                }
                                                item.setComment(comment);
                                                DBManager.getInstance(historyView.getContext()).updateValue(item);
                                                adapter.notifyItemChanged(position);
                                            }
                                        });
                                AlertDialog alertComment = commentBuilder.create();
                                alertComment.show();
                                String comment = historyItems.get(position).getComment();
                                if (comment != null) {
                                    input.setText(comment);
                                    input.setSelection(0, comment.length());
                                }
                                break;
                            case MENU_DELETE:
                                AlertDialog.Builder deleteBuilder = new AlertDialog.Builder(historyView.getContext());
                                deleteBuilder
                                        .setMessage(R.string.confirm_delete_title)
                                        .setCancelable(false)
                                        .setNegativeButton(android.R.string.cancel, new DialogInterface.OnClickListener() {
                                            @Override
                                            public void onClick(DialogInterface dialog, int which) {
                                                dialog.cancel();
                                            }
                                        })
                                        .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                                            @Override
                                            public void onClick(DialogInterface dialog, int which) {
                                                HistoryItem item = historyItems.get(position);
                                                DBManager.getInstance(historyView.getContext()).deleteValue(item.getId());
                                                historyItems.remove(position);
                                                adapter.notifyItemRemoved(position);
                                            }
                                        });
                                AlertDialog alertDelete = deleteBuilder.create();
                                alertDelete.show();
                                break;
                        }
                    }
                });
                AlertDialog alert = menuBuilder.create();
                alert.show();
            }
        }));
    }

    public void searchItems() {
        AlertDialog.Builder searchBuilder = new AlertDialog.Builder(historyView.getContext());
        final View view = getLayoutInflater().inflate(R.layout.search_panel, null);
        searchBuilder
                .setMessage(R.string.search_title)
                .setCancelable(false)
                .setView(view)
                .setNegativeButton(android.R.string.cancel, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.cancel();
                    }
                })
                .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        EditText commentText = (EditText) view.findViewById(R.id.search_comment);
                        String comment = commentText.getText().toString();
                        EditText dateText = (EditText) view.findViewById(R.id.search_date);
                        String date = dateText.getText().toString();
                        historyItems = DBManager.getInstance(historyView.getContext()).search(comment, date);
                        adapter = new HistoryAdapter(historyItems);
                        historyView.setAdapter(adapter);
                    }
                });
        AlertDialog alert = searchBuilder.create();
        alert.show();
    }

    public void clearItems() {
        AlertDialog.Builder clearBuilder = new AlertDialog.Builder(historyView.getContext());
        clearBuilder
                .setMessage(R.string.confirm_clear_title)
                .setCancelable(false)
                .setNegativeButton(android.R.string.cancel, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.cancel();
                    }
                })
                .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        DBManager.getInstance(historyView.getContext()).deleteAll();
                        historyItems.clear();
                        adapter.notifyDataSetChanged();
                    }
                });
        AlertDialog alert = clearBuilder.create();
        alert.show();
    }

    public void OnDateTextClick(View view) {
        final EditText dateText = (EditText) view.findViewById(R.id.search_date);
        final Calendar calendar = Calendar.getInstance();

        DatePickerDialog.OnDateSetListener listener = new DatePickerDialog.OnDateSetListener() {
            @Override
            public void onDateSet(DatePicker view, int year, int month, int dayOfMonth) {
                calendar.set(year, month, dayOfMonth);
                SimpleDateFormat sdf = new SimpleDateFormat(DATE_FORMAT, Locale.US);
                String date = sdf.format(calendar.getTime());
                dateText.setText(date);
            }
        };

        DatePickerDialog picker = new DatePickerDialog(view.getContext(), listener, calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH), calendar.get(Calendar.DAY_OF_MONTH));
        picker.show();
    }
}

File: app/src/main/java/com/goloveschenko/calculator/activity/HistoryActivity.kt
package com.goloveschenko.calculator.activity

import android.app.Activity
import android.app.DatePickerDialog
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.text.InputType
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.EditText
import com.goloveschenko.calculator.R
import com.goloveschenko.calculator.dao.manager.DBManager
import com.goloveschenko.calculator.entity.HistoryItem
import com.goloveschenko.calculator.tools.HistoryAdapter
import com.goloveschenko.calculator.tools.RecyclerItemClickListener
import kotlinx.android.synthetic.main.activity_history.*
import java.text.SimpleDateFormat
import java.util.*
import kotlin.collections.ArrayList

class HistoryActivity : AppCompatActivity() {
    private var historyItems: MutableList<HistoryItem> = ArrayList()
    private var adapter: HistoryAdapter = HistoryAdapter(historyItems)

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.history_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_search -> {
                searchItems()
                return true
            }
            R.id.menu_clear -> {
                clearItems()
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_history)

        historyView.layoutManager = LinearLayoutManager(this)

        historyItems = DBManager.getInstance(historyView.context).selectValues()
        adapter = HistoryAdapter(historyItems)
        historyView.adapter = adapter

        historyView.addOnItemTouchListener(RecyclerItemClickListener(historyView.context, historyView, object : RecyclerItemClickListener.OnItemClickListener {
            override fun onItemClick(view: View, position: Int) {
                val res = resources
                val menuItems = res.getStringArray(R.array.history_item_input_menu)
                val item = historyItems[position]
                val menuBuilder = AlertDialog.Builder(view.context)
                menuBuilder.setTitle(R.string.history_input_title)
                menuBuilder.setItems(menuItems) { dialog, which ->
                    var result = ""
                    when (which) {
                        MENU_INPUT_RESULT -> result = item.result
                        MENU_INPUT_EXPRESSION -> result = item.expression
                    }
                    val intent = Intent()
                    intent.putExtra(EXTRA_EXPRESSION_RESULT, result)
                    setResult(Activity.RESULT_OK, intent)
                    finish()
                }
                val alert = menuBuilder.create()
                alert.show()
            }

            override fun onLongItemClick(view: View, position: Int) {
                val res = resources
                val menuItems = res.getStringArray(R.array.history_item_menu)
                val menuBuilder = AlertDialog.Builder(view.context)
                menuBuilder.setTitle(R.string.history_item_alert_title)
                menuBuilder.setItems(menuItems) { dialog, which ->
                    when (which) {
                        MENU_COMMENT -> {
                            val commentBuilder = AlertDialog.Builder(historyView.context)
                            val input = EditText(historyView.context)
                            input.inputType = InputType.TYPE_CLASS_TEXT
                            commentBuilder
                                    .setCancelable(false)
                                    .setTitle(R.string.input_comment_title)
                                    .setView(input)
                                    .setNegativeButton(android.R.string.cancel) { dialog, which -> dialog.cancel() }
                                    .setPositiveButton(android.R.string.ok) { dialog, which ->
                                        val item = historyItems[position]
                                        val comment = input.text.toString()
                                        item.comment = comment
                                        DBManager.getInstance(historyView.context).updateValue(item)
                                        adapter.notifyItemChanged(position)
                                    }
                            val alertComment = commentBuilder.create()
                            alertComment.show()
                            val comment = historyItems[position].comment
                            if (comment != null) {
                                input.setText(comment)
                                input.setSelection(0, comment.length)
                            }
                        }
                        MENU_DELETE -> {
                            val deleteBuilder = AlertDialog.Builder(historyView.context)
                            deleteBuilder
                                    .setMessage(R.string.confirm_delete_title)
                                    .setCancelable(false)
                                    .setNegativeButton(android.R.string.cancel) { dialog, which -> dialog.cancel() }
                                    .setPositiveButton(android.R.string.ok) { dialog, which ->
                                        val item = historyItems[position]
                                        DBManager.getInstance(historyView.context).deleteValue(item.id!!)
                                        historyItems.removeAt(position)
                                        adapter.notifyItemRemoved(position)
                                    }
                            val alertDelete = deleteBuilder.create()
                            alertDelete.show()
                        }
                    }
                }
                val alert = menuBuilder.create()
                alert.show()
            }
        }))
    }

    fun searchItems() {
        val searchBuilder = AlertDialog.Builder(historyView.context)
        val view = layoutInflater.inflate(R.layout.search_panel, null)
        searchBuilder
                .setMessage(R.string.search_title)
                .setCancelable(false)
                .setView(view)
                .setNegativeButton(android.R.string.cancel) { dialog, which -> dialog.cancel() }
                .setPositiveButton(android.R.string.ok) { dialog, which ->
                    val commentText = view.findViewById(R.id.search_comment) as EditText
                    val comment = commentText.text.toString()
                    val dateText = view.findViewById(R.id.search_date) as EditText
                    val date = dateText.text.toString()
                    historyItems = DBManager.getInstance(historyView.context).search(comment, date)
                    adapter = HistoryAdapter(historyItems)
                    historyView.adapter = adapter
                }
        val alert = searchBuilder.create()
        alert.show()
    }

    fun clearItems() {
        val clearBuilder = AlertDialog.Builder(historyView.context)
        clearBuilder
                .setMessage(R.string.confirm_clear_title)
                .setCancelable(false)
                .setNegativeButton(android.R.string.cancel) { dialog, which -> dialog.cancel() }
                .setPositiveButton(android.R.string.ok) { dialog, which ->
                    DBManager.getInstance(historyView.context).deleteAll()
                    historyItems.clear()
                    adapter.notifyDataSetChanged()
                }
        val alert = clearBuilder.create()
        alert.show()
    }

    fun OnDateTextClick(view: View) {
        val dateText = view.findViewById(R.id.search_date) as EditText
        val calendar = Calendar.getInstance()

        val listener = DatePickerDialog.OnDateSetListener { view, year, month, dayOfMonth ->
            calendar.set(year, month, dayOfMonth)
            val sdf = SimpleDateFormat(DATE_FORMAT, Locale.US)
            val date = sdf.format(calendar.time)
            dateText.setText(date)
        }

        val picker = DatePickerDialog(view.context, listener, calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH), calendar.get(Calendar.DAY_OF_MONTH))
        picker.show()
    }

    companion object {
        val DATE_FORMAT = "dd MMM yyyy"
        val EXTRA_EXPRESSION_RESULT = "result"

        val MENU_INPUT_RESULT = 0
        val MENU_INPUT_EXPRESSION = 1
        val MENU_COMMENT = 0
        val MENU_DELETE = 1
    }
}
